### ISSUE TITLE
Html script closing-tag not detected when occurs immediately after 2 slashes

### ISSUE THREAD

**Issue by SDP190** [Microsoft/vscode-textmate#49]
Hi, I found a bug, with the html grammar in VS Code, when you place a closing script-tag in the same line immediately after 2 fwd slashes, as:
`<script>  //</script>`
the grammar will not detect the closing tag, and will display it with wrong color, see images.

(I'm not sure if the issue is related only to [vscode-textmate](https://github.com/Microsoft/vscode-textmate), or it belongs to [textmate/html.tmbundle)](https://github.com/textmate/html.tmbundle).

VS Code:
![sample](https://user-images.githubusercontent.com/16751213/27235025-2fea2a02-528d-11e7-9132-4acd6ab80a2c.JPG)

Visual Studio 2015:
![capture](https://user-images.githubusercontent.com/16751213/27235056-4b630e20-528d-11e7-9479-567e0921b026.JPG)

---
- VSCode Version: Code 1.13.1 (379d2efb5539b09112c793d3d9a413017d736f89, 2017-06-14T18:21:47.485Z)
- OS Version: Windows_NT ia32 10.0.15063

---

Steps to Reproduce:

1. Open any html document with VSCode.
2. Insert a script element in the document, and place the script's closing-tag immediately after 2 fwd slashes.

**Comment by alexandrudima** [comment]
I think this is not a vscode-textmate issue. The problem here is that the html grammar includes `source.js` which is HTML agnostic and "eats up" all characters of a line comment to the end of the line for example. IMHO we would need to have a special `source.html-embedded.js` or something which is always looking for `</script>`. AFAIK there are no other workarounds for an embedded TM grammar.

**Comment by SDP190** [comment]
I posted this issue [here](https://github.com/textmate/html.tmbundle/issues/72) on textmate's repository, here is what @infininight commented:
This does not happen in TextMate with the current bundle. You will need to talk with the VSCode devs to see if it is up-to-date or a difference in the parsing engines.

**Comment by alexandrudima** [comment]
Yes, this is a VS Code only problem that occurs only with the combination of HTML + JS grammars that are shipped with VS Code. The problem does not occur with the HTML + JS grammars that are shipped with TextMate.

In other words, it is not a problem with the vscode-textmate (a TM grammar interpreter), but with the grammars and their usage together.

**Comment by SDP190** [comment]
Still not fixed, WOW!

**Comment by infininight** [comment]
So an explanation of this issue so it can better be tracked down:

In Javascript generally `//` marks the rest of the line as a comment, in a pure JavaScript document `</script>` tags could be present so ending on them isn't proper. To cover for this eventuality in the HTML grammar we include a special comment rule that is aware of the closing tag and include it higher than the JavaScript grammar include so it should match before the generic comment rule from the JS grammar itself.

So the HTML grammar portion of this should be the same in VSCode as far as I know. So I suspect the issue would be on the JavaScript side of the grammar with it's rule matching instead of the special case comment rule. I have very little knowledge of how VSCode is structured so I don't know how JavaScript is handled so further investigations on that side will have to be done by someone else.

There is a wider issue of blocks in general being able to break the `</script>` tag closure so this may be something we investigate in the future on the TextMate side as I have just introduced the initial support of being able to highlight JavaScript code in the event handler attributes and this will require a specialized grammar variant.

**Comment by msftrncs** [comment]
@SDP190 , the HTML.TMLANGUAGE.JSON file needs the following fix, making line 1850 (from VS Code 127.2) read as:
```json
	"end": "(?=</script)|(?!\\G)",
```

**Comment by infininight** [comment]
@msftclas I have not tested this in VSCode but I don't see why that would work, unless VSCode doesn't support the `\G` flag which is a non-standard addition we've added in TextMate.

In TextMate `\G` matches the end of the begin match. So what your end should do is end when it sees `</script` as a look-ahead or as soon as `\G` is not true. Since `\G` should be the end of the begin that means as soon as we advance the parse one character, wether we match anything or not. This isn't the desired outcome as the comment rule would end after a single space character.

Now it may be that VSCode doesn't support `\G` in the same manner which is totally possible. In this case the behavior would be undefined and you might be seeing the right behavior for the wrong reasons.

Since your investigating: `(?=</script)` should already be ending the rule when necessary, see if you can track down why it is not matching.

**Comment by msftrncs** [comment]
The reason it was not matching, is because the search term was `//</script>` at the start, it matched `//` to begin, and was waiting for `(?!\G)` to end, while waiting at the `<`, but the `(?!\G)` doesn't 'HIT' until the next character, and with nothing else to match in that block, the cursor advanced to the `/script>` point, making any possibility to match `</script` impossible.  So by giving the end an alternate of the actual term we're looking for `</script` (using a lookahead) we can short-circuit the begin-end rule.

I wonder if textmate answers the (?!\G) without advancing the cursor?  Like maybe its 'looking ahead' to see if the next position would not match the anchor?

**Comment by msftrncs** [comment]
Okay, the difference I can see with VSCode-TextMate is that the \G anchor point when it hits the `(?=</script)|\n` is as you described, pointing to the `<` and there is where it stays when that end-match completes, so also when the previous BeginEnd end match is tested for, `(?!\G)`.  At this point, we're stuck in this simple rule.  The `(?!\G)` can find a match at the next character location, and the only other rule `//` cannot find any matches on the rest of the line, so the cursor is advanced to the next character, the end rule was given the match, and the rule is popped.  Now we're looking for a `(?=</(?i:script))` as the end match for the previous rule, but the engine sees `/script>`, no match.

I suspect that TextMate's handling of the anchor position (for matching \G) is different when it pop's from a rule, than is VSCode-TextMate's.  I know from wondering through some test runs that VSCode's behavior is that if the cursor hasn't advanced past the end of the last BEGIN rule's match, \G is given the match, and that would make a big difference if TextMate treats it, after ENDing a rule, it restores the previous rule's BEGIN match point for the anchor comparison.  I could see the syntax (without my proposed change) working in that situation.

Here is a debug of vscode (trimmed) ( x,{y} is cursor position {anchor position}, |x| only the active part of the line is shown, and matched: x / y is matched cursor position / final cursor position. )
```
@@scanNext 0,{0}: |//</script>\n|
  scanning for
   - -1: (?=</script)|(?!\G)
   - 1043: //
matched: 0 / 2
  pushing BeginEndRule#1043 @ html.tmLanguage.json:1852 - //
  token: |//|
      * text.html.basic
      * meta.embedded.block.html
      * source.js
      * comment.line.double-slash.js
      * punctuation.definition.comment.js

@@scanNext 2,{2}: |</script>\n|
  scanning for
   - -1: (?=</script)|\n
matched: 2 / 2
  popping BeginEndRule#1043 @ html.tmLanguage.json:1852 - (?=</script)|\n

@@scanNext 2,{2}: |</script>\n|
  scanning for
   - -1: (?!\G)
   - 1043: //
matched: 3 / 3
  popping BeginEndRule#1041 @ html.tmLanguage.json:1843 - (?!\G)
  token: |<|
      * text.html.basic
      * meta.embedded.block.html
      * source.js
```
Notice how in the last segment, the end rule fired, but at position 3, and passed over the `<`.  So this is the next match vscode attempts:
```
@@scanNext 3,{2}: |/script>\n|
  scanning for
   - -1: (?=</(?i:script))
   - 1041: (^[ \t]+)?(?=//)
   … (400 lines of javascript rules removed)
matched: 3 / 4
  matched MatchRule#597 @ JavaScript.tmLanguage.json:2924 - %|\*|/|-|\+
  token: |/|
      * text.html.basic
      * meta.embedded.block.html
      * source.js
      * keyword.operator.arithmetic.js
```
In this case, I am betting that after the BeginEndRule#1043 is popped in TextMate's engine, it reverts the anchor match condition to a value representing that no anchor exists on the current line.  I could see this being the correct behavior.  However, I would prefer something in documentation that tells tmlanguage developers that is the case, that the anchor \G only matches at the end of the BEGIN rule for the current active BeginEnd rule.

**Comment by msftrncs** [comment]
Note, the {anchor position} is my addition in my fork. so I could diagnose an anchor problem.

**Comment by infininight** [comment]
@msftrncs Sorry missed your followup question but got pointed to this from a Pug related question.

In this example when we exit the comment rule by `(?=</script)|\n` this would invalidate that rules `\G` match. I could check to see if the parent rules `\G` is restored but I can't think of any time it would be relevant. The two uses of `\G` would be make sure we are still beside the begin or ensure we aren't. Entering and leaving a pattern array without consuming any characters would be the only way we would need to rematch a previous \G match but this would cause an infinite loop anyway so not really relevant.

So I think any issues can be solved by making sure the anchor is cleared when leaving an end and either restoring the parent rules state if any or just removing it entirely. Or to answer your closing question, yes the `\G` anchor only matches the begin of the current begin/end rule or begin/while rule.

cc Microsoft/vscode-textmate#82

**Comment by alexandrudima** [comment]
Thank you @msftrncs , this issue is also fixed by #85
